### ISSUE TITLE
[question] Can there be any benefits of fusing qkv-projection into attention (if possible at all)

### ISSUE THREAD

**Issue by vadimkantorov** [HazyResearch/flash-attention#206]
The reason I'm asking is that in the repo/paper https://github.com/NVlabs/tiny-cuda-nn they say that for small inner dim (e.g. 64), the weight matrix can fit into shared memory. per-head-dim = 64 is quite common for transformers, so I'm wondering if qkv projection can then be fused with the rest of the attention

**Comment by tridao** [comment]
1. The QKV projection matmul is fast already, it's bottlenecked by compute and not memory, so I don't think fusing it would improve speed. Kernel fusion decreases memory reads/writes, so it's helpful for operations that are bounded by memory reads/writes.
2. It might be hard to fuse the projection, since the weight matrix is quite large. Let's take GPT-125m, which has hidden dimension 768. Then the projection matrix for each head is 768 x 64, which takes 96KB if we load the bf16 weight to SRAM. This is already more than half of the SRAM available for a threadblock on A100 (163KB). And this is just the weight matrix for the Q projection, involving K and V projection would be even harder.

**Comment by vadimkantorov** [comment]
And same for Feed-Forward parts of transformers, right? The embedding dim there is quite large, so these MLP's are probably not very fusable, right?

**Comment by vadimkantorov** [comment]
Yeah, I guess maybe your works on butterfly matrices and other structured parameterizations can be interesting in this regard! :) e.g. does butterfly matrices here make sense? if you integrate their fused projectoin variants in flash-attention repo, it could lead to popularization of these structured representations :)

**Comment by tridao** [comment]
Yes, I've been thinking about that direction. It's interesting for the reason you said, I just haven't had the bandwidth to work on it.

**Comment by vadimkantorov** [comment]
I guess I'll close this issue. Thanks for your thoughts!

**Comment by tridao** [comment]
Thanks! I talked with Hao and my understanding is that their technique is used to reduce peak memory.
